### PR TITLE
Lazy save improvement

### PR DESCRIPTION
**Pull request by ericpre**
### Progress of the PR
- [x] Add option to close file when saving lazy signal to hspy format,
- [x] add option not to write dataset, which can useful when saving large signal (hspy/zspy format only),
- [x] fix issue with latest `h5py` version (3.5)
- [x] update docstring,
- [x] update user guide,
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
import numpy as np

s = hs.signals.Signal1D(np.arange(1000).reshape(10, 100))

fname = 'test.hspy'
s.save(fname, overwrite=True)

s2 = hs.load(fname, lazy=True, mode='a')
s2.axes_manager[-1].scale = 0.1
# write everything, except the dataset itself
s2.save(fname, overwrite=True, write_dataset=False)

s3 = hs.load(fname)
print(s3.axes_manager[-1].scale)
```

### CODE REVIEW

File: doc/user_guide/io.rst
Comment (jlaehne) [inline doc/user_guide/io.rst]: closed

File: hyperspy/io.py
Comment (jlaehne) [inline hyperspy/io.py]: to write (not the write)

File: hyperspy/io_plugins/hspy.py
Comment (jlaehne) [inline hyperspy/io_plugins/hspy.py]: to write

**Review by jlaehne — COMMENTED** [review]
some typos

**Comment by ericpre** [comment]
I will come back to this after #2825 is merged.

File: hyperspy/io_plugins/_hierarchical.py
Comment (ericpre) [inline hyperspy/io_plugins/_hierarchical.py:576]: This is never used and it breaks with latest version of `h5py` (3.5).

File: hyperspy/io_plugins/zspy.py
Comment (ericpre) [inline hyperspy/io_plugins/zspy.py:150]: This was actually not working as expected when opening a file which has been saved with a store different than the `NestedDirectoryStore` or `DirectoryStore`, because `zarr` is initialising the correct store based on the extension (https://github.com/zarr-developers/zarr-python/blob/523dbb85cbad5a5fa3b74a2ec81ec7ec63123b64/zarr/creation.py#L160-L178). For this reason, I remove it here.

File: hyperspy/io_plugins/zspy.py
Comment (ericpre) [inline hyperspy/io_plugins/zspy.py:125]: Using `zarr` group private attribute is asking for trouble!
@CSSFrancis, in #2825, you mentioned that this was necessary in some cases? I have extended the test coverage and this seems to be working fine and I think that this use of the API is supported, so if this doesn't work, it could be a bug somewhere else.
Can you please check?

**Review by ericpre — COMMENTED** [review]
The test suite is failing on python 3.6 and I think that it is not worth trying to get it to work because python 3.6 is not supported anymore and we should drop it in the next minor release.

**Comment by ericpre** [comment]
@CSSFrancis, would you like to review this PR? Thanks!

File: hyperspy/io_plugins/_hierarchical.py
Comment (CSSFrancis) [reply hyperspy/io_plugins/_hierarchical.py:576]: That is good to know.  I figured that was left over from previous version but it wasn't tested so I didn't know if there was a case where it was important.

**Comment by CSSFrancis** [comment]
> @CSSFrancis, would you like to review this PR? Thanks!

Yea let me see if I can get to this today or tomorrow

File: hyperspy/io_plugins/zspy.py
Comment (CSSFrancis) [inline hyperspy/io_plugins/zspy.py]:
```suggestion
            da.store(data, dset, lock=False)
```

File: hyperspy/io_plugins/zspy.py
Comment (CSSFrancis) [reply hyperspy/io_plugins/zspy.py:125]: So the issue arises when using a dask.distributed scheduler. The solution is to set lock = False.  I didn't realize this when I originally wrote the code.  It might be worth writing tests in the future to test dask.distrubted compatibility if we are planning to fully support it (Which I would recommend)

Comment (CSSFrancis) [reply hyperspy/io_plugins/zspy.py:125]: It seems like the pickling is intending to be thread safe so it throws an error if lock is set to True

File: hyperspy/io_plugins/zspy.py
Comment (CSSFrancis) [reply hyperspy/io_plugins/zspy.py:150]: Ahh, very true.  Removing the write to storage argument should be fine.

File: hyperspy/tests/io/test_zspy.py
Comment (CSSFrancis) [inline hyperspy/tests/io/test_zspy.py]: This should not be deleted I think.  You can remove the `write_to_storage=True` but this test shows if passing a MutableMapping object works for saving the data.

**Review by CSSFrancis — CHANGES_REQUESTED** [review]
@ericpre Looks like this cleans up a lot of the little things that I wasn't terribly confident about and reduces the code complexity a fair bit! Thanks for doing this. 

One note (as I said above) is that the `store` function doesn't play well with dask.distrubted unless lock=False.  It might be worth adding in some tests for compatibility there if the goal is to fully integrate with dask, but I am not sure how easy deploying something like that on a testing environment is. 

I assume that replacing `create_group` with `require_group` is preferable because if won't throw an error if the group already exists?

Other than that, one of the tests which tested passing a Mutable Mapping object as the file name was deleted and should be covered.

**Comment by ericpre** [comment]
Thanks for the review!

> One note (as I said above) is that the store function doesn't play well with dask.distrubted unless lock=False. It might be worth adding in some tests for compatibility there if the goal is to fully integrate with dask, but I am not sure how easy deploying something like that on a testing environment is.

Yes, this is what dask is doing too: https://github.com/dask/dask/blob/62f96e7edb27cb94eca49d5ac2f5639fb8cca44a/dask/array/core.py#L3388

> I assume that replacing create_group with require_group is preferable because if won't throw an error if the group already exists?

Yes, this is correct, same for `h5py` and [`zarr`](https://zarr.readthedocs.io/en/stable/api/hierarchy.html?highlight=require_group#zarr.hierarchy.Group.require_group).

> Other than that, one of the tests which tested passing a Mutable Mapping object as the file name was deleted and should be covered.

As mentioned in https://github.com/hyperspy/hyperspy/pull/2797#discussion_r733619173, this was not working correctly, but I had a look at it again, one way to get it to work is to pass store of the same in the `load` function.

It should be all good now!

File: hyperspy/io_plugins/zspy.py
Comment (ericpre) [reply hyperspy/io_plugins/zspy.py:125]: > So the issue arises when using a dask.distributed scheduler. The solution is to set lock = False. I didn't realize this when I originally wrote the code. It might be worth writing tests in the future to test dask.distrubted compatibility if we are planning to fully support it (Which I would recommend)

Yes, it would be good to have some tests using `dask.distributed.Client`. I don't know how well it will play in CI and I had a quick look at how dask are testing it, and I didn't understood straightaway, so I would leave it for another PR! :)

File: hyperspy/io_plugins/hspy.py
Comment (ericpre) [inline hyperspy/io_plugins/hspy.py:267]: `overwrite_dataset` is actually part of the public API, so I put back here... it was breaking kikuchipy, as pointed out by https://github.com/hyperspy/hyperspy/runs/3973795422?check_suite_focus=true!

**Comment by ericpre** [comment]
Running the integration test suite show that it was breaking the public API... see https://github.com/hyperspy/hyperspy/pull/2797#discussion_r734456427. This is fixed in caf6a01.

@CSSFrancis, does it look good to you?

**Comment by CSSFrancis** [comment]
> @CSSFrancis, does it look good to you?

Yea with the changes it looks good! Thanks for cleaning this stuff up.

File: hyperspy/io_plugins/hspy.py
Comment (hakonanes) [reply hyperspy/io_plugins/hspy.py:267]: Very good catch, thanks a lot.

**Comment by ericpre** [comment]
@jlaehne, do you want to do another review of this PR?

File: hyperspy/io.py
Comment (jlaehne) [inline hyperspy/io.py]: In the user guide it says these extra arguments are only relevant for `zarr` format, here it says they are only relevant for `hspy` format. Also some details in the description differ between here and io.rst.

**Comment by jlaehne** [comment]
> @jlaehne, do you want to do another review of this PR?

I sent a PR correcting some typos. There are still some codecov warnings (mostly about untested exceptions). Otherwise nothing that caught my eye.

File: hyperspy/io_plugins/_hierarchical.py
Comment (jlaehne) [inline hyperspy/io_plugins/_hierarchical.py]: sorted or stored ?

File: hyperspy/io.py
Comment (ericpre) [reply hyperspy/io.py]: Yes, this docstring is for `save`, not `load`! :facepalm:

**Comment by ericpre** [comment]
Thanks @jlaehne, I increased the coverage and fixed the docstring.

File: doc/user_guide/io.rst
Comment (magnunor) [inline doc/user_guide/io.rst]: Typo: should be integer

File: hyperspy/io_plugins/_hierarchical.py
Comment (magnunor) [inline hyperspy/io_plugins/_hierarchical.py]: signal_axes : {None, iterable of ints}

File: hyperspy/io_plugins/_hierarchical.py
Comment (magnunor) [inline hyperspy/io_plugins/_hierarchical.py]: target_size : int

File: hyperspy/io_plugins/_hierarchical.py
Comment (magnunor) [inline hyperspy/io_plugins/_hierarchical.py]:
```suggestion
    An object used to simplify and organize the process for writing a
```

File: hyperspy/io_plugins/_hierarchical.py
Comment (magnunor) [inline hyperspy/io_plugins/_hierarchical.py]:
```suggestion
    Hierarchical signal, such as hspy/zspy format.
```

File: doc/user_guide/io.rst
Comment (magnunor) [inline doc/user_guide/io.rst]:
```suggestion
  a compressor can be passed to the save function to compress the data efficiently. The default
```

**Review by magnunor — COMMENTED** [review]
I tested this on my own computer, and I'm getting an error:

```python
import dask.array as da
import hyperspy.api as hs

s = hs.signals.Signal1D(da.random.random((10, 20, 50))).as_lazy()
s.save("test_dataset.hspy")
```
Giving the error:
```python
hyperspy/io_plugins/_hierarchical.py in overwrite_dataset(cls, group, data, key, signal_axes, chunks, **kwds)
    574                     # we delete the old one and create new in the next loop run
    575                     del group[key]
--> 576         if dset == data:
    577             # just a reference to already created thing
    578             pass

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

I do have a bit of a strange mix of package versions, so it could be caused by that.

**Comment by ericpre** [comment]
> I tested this on my own computer, and I'm getting an error:
> 
> ```python
> import dask.array as da
> import hyperspy.api as hs
> 
> s = hs.signals.Signal1D(da.random.random((10, 20, 50))).as_lazy()
> s.save("test_dataset.hspy")
> ```
> 
> Giving the error:
> 
> ```python
> hyperspy/io_plugins/_hierarchical.py in overwrite_dataset(cls, group, data, key, signal_axes, chunks, **kwds)
>     574                     # we delete the old one and create new in the next loop run
>     575                     del group[key]
> --> 576         if dset == data:
>     577             # just a reference to already created thing
>     578             pass
> 
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
> ```
> 
> I do have a bit of a strange mix of package versions, so it could be caused by that.

Thanks @magnunor! It seems that you are on a wrong branch, the code at line 576 is different in this branch:
https://github.com/ericpre/hyperspy/blob/9c5bf6fa1550867eee8a586ad4a8d163d38f4f2a/hyperspy/io_plugins/_hierarchical.py#L576

**Comment by magnunor** [comment]
> It seems that you are on a wrong branch, the code at line 576 is different in this branch:

Indeed! Seems like `git` had changed how it handled pulling branches from a remote, so this,

```
git checkout -b ericpre-lazy_save_improvement RELEASE_next_minor
git pull https://github.com/ericpre/hyperspy.git lazy_save_improvement
```
failed.

File: hyperspy/io_plugins/_hierarchical.py
Comment (magnunor) [inline hyperspy/io_plugins/_hierarchical.py]:
```suggestion
        Initializes a general reader for hierarchical signals.
```

File: hyperspy/io_plugins/hspy.py
Comment (magnunor) [inline hyperspy/io_plugins/hspy.py]:
```suggestion
    An object used to simplify and organize the process for
```

File: doc/user_guide/io.rst
Comment (magnunor) [inline doc/user_guide/io.rst:417]: Is this sentence correct? Or at least ambiguous?

I'm guessing it should say that using compression can cause file saving and loading to be much slower.

**Review by magnunor — COMMENTED** [review]
I got the previous issue sorted out, and tested the functionality a little bit. Some observations:

- `write_dataset=False` for updating both `hspy` and `zspy` files works fine, for both `axes_manager` and `metadata`
- `zspy` does not require loading the data with any specific `mode`. As opposed to `hspy`, which requires `mode="a"`
- The filename can be dropped when doing `s.save(overwrite=True, write_dataset=False)`
- Loading via `zspy` is much faster, both using `DirectoryStore` and `LMDBStore`. For a 512 x 512 x 256 x 256 dataset, with uint16, it took about 20 seconds. The same dataset took about 100 seconds with `hspy`

This is probably unrelated, but doing:

```python
s = hs.load("test_data.hspy", lazy=True)
s.compute()
```
used two times as much memory (64 GB), as opposed to `s = hs.load("test_data.hspy", lazy=False)`, which used 32 GB.

This is using the most recent version of `dask`, `2021.10.0`

File: doc/user_guide/io.rst
Comment (ericpre) [reply doc/user_guide/io.rst:417]: Yes, this is correct in many cases when the IO time is balanced against with the CPU time - most of the time, CPU are fast enough and compressor are efficient enough.

**Comment by ericpre** [comment]
> This is probably unrelated, but doing:
> 
> ```python
> s = hs.load("test_data.hspy", lazy=True)
> s.compute()
> ```
> 
> used two times as much memory (64 GB), as opposed to `s = hs.load("test_data.hspy", lazy=False)`, which used 32 GB.
> 
> This is using the most recent version of `dask`, `2021.10.0`

Yes, this is not related to this PR - at least I don't see how it can be and I would expect that the same happen with other branches. Do you observe the same memory usage with `zspy`?

**Comment by ericpre** [comment]
Thanks @CSSFrancis, @jlaehne and @magnunor for the reviews, I will merge this PR to be able to rebase https://github.com/hyperspy/hyperspy/pull/2839 and fix CI, so that it is easier to work on #2842 and co.

**Comment by magnunor** [comment]
> Yes, this is not related to this PR - at least I don't see how it can be and I would expect that the same happen with other branches. Do you observe the same memory usage with zspy?

Yes, this was also with `zspy`.

**Pull request merged**
